Dapp rpc whitelist

sebastiandelgado opened the pull request:
This PR replaces https://github.com/status-im/status-go/pull/922

It is identical to #922 except that it is on a branch in status-im/status-go instead of being on a branch in my fork sebastiandelgado/status-go

Commit:
Add isExternal param to CallRaw for rpc whitelist

Add the isExternal param in order to implement the external RPC call whitelist detailed in #912

adambabik commented on geth/rpc/call_raw.go:
```diff
@@ -20,10 +25,36 @@ const (
 
+// All methods from the namespaces in this whitelist are
+// allowed in external RPCs
+var externalRPCNamespaceWhitelist = []string{
```
This is kind of duplication of `APIModules` from `defaults.go`. We transform `APIModules` to the list.

Also, @mandrigin shouldn't we remove `shh` from `APIModules`?

mandrigin replied:
my idea was to have a separate PR for that

adambabik replied:
That's fine. Let's make sure #912 is not closed when merging this PR then.

Still it's a duplication. I would do
```go
var externalRPCNamespaceWhitelist = strings.Split(params.APIModules, ",")
```

mandrigin replied:
I'd rather have it another way around :) Or, frankly, we can get rid of this whitelist in config whatsoever.

adambabik replied:
I don't see how it's gonna work without `APIModules`. `APIModules` must be passed to the node config, otherwise, even if something will be whitelisted here, it will never reach the RPC server.

sebastiandelgado commented:
Started using APIModules for the namespace whitelist. Still need to figure out whats wrong with travis ci that has this PR absent from https://travis-ci.org/status-im/status-go/branches

<blockquote><div><strong><a href="https://travis-ci.org/status-im/status-go/branches">Travis CI - Test and Deploy Your Code with Confidence</a></strong></div></blockquote>

Commit:
External RPC method whitelist
Addresses #912

Create a whitelist that prevents external RPC requests (from dapps) from using methods like shh.setMinPoW. The approach here was to pass down an 'external' flag into the context.Context of call raw, since its a request-scoped variable

pedropombeiro commented:
@sebastiandelgado I noticed that you're not part of the status-im org. Maybe that's why your PRs don't build in Travis CI? (reaching out to Stef)

adambabik reviewed (approved):
LGTM

Do we have an issue to remove `shh` from `APIModules`?

sebastiandelgado commented:
@adambabik as agreed offline, will submit a separate PR to remove `shh` from APIModules once this one gets merged

divan reviewed (commented):
The code is good, however, I see how https://github.com/status-im/status-go/pull/825 started to bring more complexity in the RPC subsystem.

We now have not only public/private RPC call methods, but also exported/unexported, and I'm not sure it's easy to tell whether they are related. ("exported" assumed to be able to call only "public" RPC methods, rights?)

The only reason why `CallPrivateRPC` exist is to allow `status-react` to get the output of the `admin_peers` (https://github.com/status-im/status-react/issues/3888). I propose  to ditch `CallPrivateRPC` and use our "old" API between `status-react` and `status-go`: just add new binding `AdminPeers` – it's guaranteed that dApps won't have access to it.

That will simplify the code a lot, keeping RPC package as simple as possible, while still allowing status-go part to control who has access to what.

cc @adambabik @dshulyak

adambabik commented:
> The only reason why CallPrivateRPC exist is to allow status-react to get the output of the admin_peers.

It's not only that. For example, we need to disable `shh` APIs for DApps (otherwise it's possible to block all communication by any DApp or retrieve private keys) but keep it available for status-react. We would need to create a lot of bindings if we get rid of `CallPrivateRPC`. For me, it's pretty clear what can be called from `CallPrivateRPC` and `CallRPC`. Personally, I favour `CallPrivateRPC` over writing individual bindings.

The problem we are trying to solve here, I guess, is a situation when we can't block the whole namespace. In this case, it's `personal`. However, I don't see the whole picture why we need to do it at this point of time. Currently, `personal` is not part of `APIModules`, so it's not available for DApps, unless we have custom handlers and they are available, but then why do we need to whitelist these handlers?

Original issue (#912) only mentions about `shh` APIs and these can be disabled in `APIModules`.

sebastiandelgado commented:
@mandrigin can you chime in on this?

@adambabik what you mean is that if we scrap this PR and remove `shh` from `APIModules` we will have effectively a whitelist that doesn't include personal and shh? What about `personal_sign` and `personal_ecRecover`? I thought those did need to be made available to dapps (but not anything else from `personal`).

If we need `personal_sign` and `personal_ecRecover` then we need to keep the PR and make a separate one for removing shh from `APIModules` and if we don't need them, then we can easily scrap this PR and make a new one just removing shh from `APIModules`.

adambabik commented:
@sebastiandelgado 
> what you mean is that if we scrap this PR and remove shh from APIModules we will have effectively a whitelist that doesn't include personal and shh?

Correct if both `personal` and `shh` are not present in `APIModules`.

> What about personal_sign and personal_ecRecover?

We need these methods to be available but in this case, we have handlers registered for them so they are available for DApps even if we remove `personal` from `APIModules`. BTW this can't be tested with `statusd` because it exposes the native `geth` HTTP server which has no idea about our handlers and our RPC wrapper. It can be fixed though.

To sum up, I think this change is not needed now. It is a solution to a problem when we need even more control over which available methods for DApps when, for example, we need to block only some methods of a given namespace. To solve the original issue #912, it's enough to remove `shh` and `shhext` from `APIModules`.

Also, this copy should be changed:
```
// APIModules is a list of modules to expose via any type of RPC (HTTP, IPC, in-proc)
// we also expose 2 limited personal APIs by overriding them in `api/backend.go`
```
to 
```
APIModules is a list of modules to expose via HTTP and `CallRPC()` binding.
We also expose all handlers registered with `rpc.Client.RegisterHandler` to `CallRPC()` binding.
```

sebastiandelgado commented:
It seems like we can address this issue by simply removing `shh` and `shhext from APIModules`, so we're closing this PR and opening #993 to replace it.
